### ISSUE TITLE
Fully-automate merges for the simplest class of package publications PRs

### ISSUE THREAD

**Issue by shonfeder** [ocaml/opam-repository#26106]
Even in the imperfect state of our current CI results, there are some package updates that are straightforward enough that we could fully automate their publication, provided we organize our CI sufficiently. We should do so.

## Benefits

I proposed this in a maintainer meeting, and the sentiment was that the packages that currently fall into this bucket are so easy to review, that it should not be a high-priority. I think I disagree here: it is easy to underestimate the benefits of automating away small friction points, and it sets us up the processes for something we are going to need before long anyhow.

Main benefits I have in mind:

- Better publication experience for package authors making these kinds of updates.
- Less busywork for maintainers, allowing us to focus on the packages that actually need human review.
- Lays the basis of a fully automated publication pipeline, upon which we can gradually expand.

## Criteria

The current requirements for a fully automated publication would be, at least:

1. The PR upates an existing package.
2. All CI checks pass
3. The author of the PR is a known maintainer/owner of the existing package

(3) Could be addressed via something like a `.codeowners` file in the package dir, or a linting check that compares the package data with the author of the PR triggering the CI process. Either of these would amount to piggybacking on github's auth, and using the repo contents as our source for permissioning configuration. 

## Required

- [ ] https://github.com/ocurrent/opam-repo-ci/issues/309
- [ ] A way to verify that the author of the PR is an owner/maintainer of the package being updated
- [ ] A workflow that will auto-merge PRs meeting the criteria.

## Related

- #23789

--- 

Additional thoughts and considerations on the requirements etc. would be helpful. :)

**Comment by hannesm** [comment]
I'm not sure why #23789 is related.

My though is: has opam-repo-ci been reviewed and developed in a style to fail on errors? I remember there have been CI issues where it resulted in "ok" but didn't do anything. With the perspective of an attacker, if I manage to figure out something that leads to such CI confusion, and auto-merge is enabled -- I could easily take over the opam-repository. It may help to have a separate CI system, e.g. using GitHub action or likewise, that does a very small amount of checks (such as maintainer metadata, the diff is straightfotward (doesn't touch too many packages, checksums aren't changed, ...). You can see that I have some reservations with "ocurrent" and "opam-repo-ci" -- which is great to have, but at the same time a pretty big chunk of code - and I'm not aware of any design documents, neither a full system specification.

With the idea of "author/maintainer" of a package, this is still very much undefined. I worked years ago to find out "who's actually maintaining package X" by using the metadata from git. I guess this is a much more resilient approach than trying to decode the "maintainer" and "author" fields of opam -- which are very inconsistent. Turns out, in the long run, for [conex](https://github.com/hannesm/conex) we need that information about packages in any case.

**Comment by shonfeder** [comment]
Thank you for the considerations :)

> I'm not sure why https://github.com/ocaml/opam-repository/issues/23789 is related.

I guess you have a narrower topic in mind with the question in the title, "how does opam-repository scale", but the amount of manual work required is a bottleneck to scaling for a more active repo.

> opam-repo-ci been reviewed and developed in a style to fail on errors?

That is certainly the hope! I'm sure there is room for improvement.

> It may help to have a separate CI system, e.g. using GitHub action or likewise, that does a very small amount of checks

I agree that it would make sense for this kind of check to be in a simple, dedicated CI job, that was very easy to reason about.

> With the idea of "author/maintainer" of a package, this is still very much undefined.

I don't think I get this point. We have a maintainer field, no? But using something like a .codeowners file may be better for the reasons you mention. One way of thinking about this effort is as a way of giving established package actors a way to have more autonomy over package publication, when everything is on the happy path. So I think it is worth considering adding support for that in way that may be orthogonal to opam's package meta data.

**Comment by yawaramin** [comment]
Isn't checking the GitHub username good enough?

- If the package is a new version of an existing one
- And the GitHub username of the [PR author](https://github.com/orgs/community/discussions/25502) is the same as the username in the opam file's `dev-repo` field
- And the build is passing

Then auto-merge the PR?

**Comment by mseri** [comment]
It was a useful exercise to try and remember where this went wrong in the past, since this would be a conveninent automation to have. The few cases where we have experienced green ticks that should have not been merged, from what comes to mind, were: 

- not submitted by maintainer or person affiliated to the package (but I think using the latest submitter or the owner of the repo should mitigate this)
- packages with incorrect metadata:
   - e.g. meaningless or stock test tags, incorrect urls, meaningless description and/or synopsis
   - included many unnecessary dependencies – ocamlformat and/or opam-publish happened for sure, maybe others
   - all checks packages were skipped due to un-metable constraints (completely uninstallable packages in this way used to be green in the CI, not sure if this is still the case)
- packages with incorrect build setup:
   - conf packages without build test and effectively installing nothing
   - wrong makefile or dune configuration, effectively installing (and compiling) nothing
- test project with a handful line of code

**Comment by shonfeder** [comment]
> Isn't checking the GitHub username good enough?
> 
>     * If the package is a new version of an existing one
> 
>     * And the GitHub username of the [PR author](https://github.com/orgs/community/discussions/25502) is the same as the username in the opam file's `dev-repo` field
> 
>     * And the build is passing
> 
> 
> Then auto-merge the PR?

Yep, this will handle some simple cases, I think! It leaves a lot of packages out, e.g., anything from a repo not on github or under an organization vs. an individual, but I think it's a good start and covers a pretty common case.

Thanks @mseri! That's very helpful. I think some of the complications you mention here can be dealt with by improved CI checks. Many of the problems seem to only apply to initial release of new packages, and currently this doesn't propose automation for that case: so for new packages, the expectation is still that it would be manually vetted.

However, in addition to only allowing automatic merges on the happy path of *already published* packages, we could start with a white list of trusted, high-frequency publishers.

**Comment by yawaramin** [comment]
Exactly. My goal is not to solve the automated publish problem comprehensively for all possible scenarios right now. My suggestion doesn't _take anything_ from or _worsen_ the experience of publishers who are not on the 'happy path' of using GitHub or being an individual publisher. I am just trying to suggest an _improved experience_ for people on the happy path, which I believe is the majority of publishers right now.

If the auto-merge check I suggested is easy to implement, then it's a low-hanging fruit that can greatly improve the opam package publishing experience for most current publishers.

**Comment by shonfeder** [comment]
Sure. This is all consistent with the explicit aim and intent of the issue. Thanks.
